Report per-thread CPU usage in ddsperf

eboasson opened the pull request:
Along with the other statistics, print user and system CPU usage for all
threads in "ddsperf" where at least one of the two is at least 0.5%.
Currently only for macOS and Linux, because those are the only ones I
can test.

Signed-off-by: Erik Boasson <eb@ilities.com>

k0ekk0ek commented:
@eboasson, I think most of the information could have been more easily retrieved using `ddsrt_getrusage` instead. That would also make this work on Windows (and FreeRTOS if the right option is enabled) too. Maybe consider changing the code to use that instead? There's of course the problem that there's currently no code in `ddsrt` that retrieves all the threads in the current process, though you might be able to use the `thread_states` vector from `q_thread.c`. Of course you could also opt to use `ddsrt_getrusage` to retrieve the information and use small portions of the code from this pull request to fetch all the active threads. Of course those small portions can be moved to `ddsrt` too and have the feature to retrieve all threads in the process only be available for macOS and Linux for now :wink:

eboasson commented:
> I think most of the information could have been more easily retrieved using ddsrt_getrusage instead. That would also make this work on Windows (and FreeRTOS if the right option is enabled) too. Maybe consider changing the code to use that instead?

I hadn't realised that of the currently suppored operating systems, it is only Linux on which the distinction between retrieving information for the calling thread is so different from retrieving infromation for an arbitrary thread ...

I'll revise it ..

> There's of course the problem that there's currently no code in ddsrt that retrieves all the threads in the current process, though you might be able to use the thread_states vector from q_thread.c.

I'd rather avoid the thread_states array for this purpose: that would mix the various modules a little more than I like (I do agree that functionality from q_thread.c ought to move into ddsrt). That's the primary reason for using the OS primitives to check.

> Of course you could also opt to use ddsrt_getrusage to retrieve the information and use small portions of the code from this pull request to fetch all the active threads. Of course those small portions can be moved to ddsrt too and have the feature to retrieve all threads in the process only be available for macOS and Linux for now

That's the direction I will take for now.

Thanks 😃

eboasson commented:
@k0ekk0ek, I've moved the platform-specific stuff into ddsrt and added Windows support for good measure. It defines yet-another type for thread identifiers, but that is because some platforms (looking at you, Linux!) do not provide functions to translate from one to the other and back.

eboasson commented:
I added some more features ... a very simplistic "netstat" addition to ddsrt, for getting information on the network I/O as reported by the kernel. I have no intention of using for more complicated things that ddsperf, but there it is nice to not have to fiddle with netstat (on macOS) or periodically reading /proc/net/dev (on Linux) to get actual network load.

k0ekk0ek commented:
@eboasson, I really like the addition of the netstat type of functions, so kudos for that :+1:. I've started the review, but the first thing that caught my eye is that there's a number of functions in `ddsrt` that return `DDS_RETCODE_UNSUPPORTED`. Please don't do that. I suggest you write a test for that specific piece of functionality in `ddsrt/cmake` (from the top of my head) and simply specify in the header that something is not supported. `DDSRT_HAVE_FEATURE_X` is defined to 0 for platform x. I've spent to much time cleaning up the aftermath of that type of code. If you want I can write such a test for you, there's not much to it. Actually, it's less work than writing a dummy implementation.

I'll do a much more thorough review soon.

eboasson commented:
I'm not so sure I agree with you that it always makes sense to do such things with macros that wiggle their way into the application sources and (and possibly even their makefiles). You're basically pushing the ugliness into the places where it gets used — sometimes that's a good thing to do, sometimes it is not ...

But I am willing to indulge you this time. It is, after all, not a bad thing to follow an established pattern as long as there are decent reasons why that pattern is the way it is.

k0ekk0ek commented:
That's exactly why I'm considering generating a configuration header after all(?) I want to keep `ddsrt` as small as possible for embedded devices and ensure that people who want to port this project to other targets don't have to go through the hassle of implementing all kinds dummy functions that don't make sense for that particular target. But you're right, it's a fine line in some cases. e.g. strictly speaking it doesn't make sense to have a `getpid` on targets like FreeRTOS, or maybe it should always return 1... We'll see, if corner cases popup, I'm open to choose a dummy implementation over the macro solution if it makes sense.

k0ekk0ek commented on src/ddsrt/cmake/thread_list.c:
```diff
@@ -0,0 +1,18 @@
+/*
```
This is not going to work(?) as it's not in the list of files that CMake checks. You can add a `threads.c` and define `cmake_HAVE_THREADS` to always be one and `cmake_HAVE_THREAD_LIST` either `true` or `false` based on the header.

eboasson replied:
Hmm ... I'll delete the test. The ``DDSRT_HAVE_THREAD_LIST`` macro is what matters for now, and whether or not it is true has no bearing on which files need to be taken into account. So there is no need to get CMake involved for now.

k0ekk0ek commented on src/tools/ddsperf/ddsperf.c:
```diff
@@ -1380,22 +1405,60 @@ static void print_stats (dds_time_t tstart, dds_time_t tnow, dds_time_t tprev)
+       most recent value.  While not entirely correct, the nature of the process
+       is such that things should be stable, and this allows printing the stats
+       always in the same way, despite the absence of synchronization (it would
+       perhaps be nicer to */
```
Comment is not complete?

eboasson replied:
Nonsensical bit deleted

eboasson commented on src/ddsrt/src/threads/windows/threads.c:
```diff
@@ -197,6 +251,13 @@ ddsrt_thread_getname(
   return cnt;
 }
 
+/** \brief Wrap thread start routine
```
stuff happens ... a bit of explanation for the new state is definitely warranted

eboasson commented on src/ddsrt/src/threads/windows/threads.c:
```diff
@@ -241,10 +306,73 @@ ddsrt_thread_setname(
+  return n;
 }
 
+dds_return_t
```
I'd rather not, at least not while the logging code relies on ``ddsrt_thread_getname``. I don't know how expensive GetThreadDescription is, and I do know it requires dynamic memory allocation. Adding a note explaining this.

eboasson commented on README.md:
```diff
@@ -1,9 +1,12 @@
 
-Eclipse Cyclone DDS is by far the most performant and robust DDS implementation available on the
-market. Moreover, Cyclone DDS is developed completely in the open as an Eclipse IoT project
+Eclipse Cyclone DDS is a very performant and robust open-source DDS implementation.  Cyclone DDS is developed completely in the open as an Eclipse IoT project
```
I've dealt with too many people that try to convince others by playing fast and loose with the truth — there are less polite words that would perhaps better describe my sentiment regarding those methods and the people employing them, but, you know, children may be reading these comments! — and who would not be able to provide solid evidence for their claims. There have even been those that "cosmetically altered" performance figures they got from me personally (not related to Cyclone, should anyone be wondering), by simply doubling the figures to get something provably impossible, and who afterward explained that "no one ever believes the numbers anyway, so better inflate them first". It sickens me ... and that form of madness certainly did not get man onto the moon and back.

I wish I had _evidence_ of Cyclone being "the most" performant DDS implementation — but it is impossible for me to ever get that evidence, and that has nothing to do with the actual performance. The reasons are simply that there are vendors that do not allow performing and/or publishing reproducible measurements, and that occasionally one discovers other implementations one was not aware of in the first place. So the old claim is simply a boast.

There _is_ evidence that Cyclone has a very good performance in general, and there are good grounds to believe that its performance (as measured in data rate or latency) outclasses that of various other implementations, perhaps all. So "very" is a defensible, if unimaginitve, choice of word.

I'd rather people look at the numbers and decide whether it will do, than have them believe grandiose claims. And if they allow themselves to be taken in by such claims without questioning the evidence behind them, then they fully deserve the lesser quality they will end up with. There is only one way to excellence, and that is honesty.

😁

eboasson commented on examples/perfscript/throughput-test-extract:
```diff
@@ -1,59 +1,76 @@
 #!/usr/bin/perl -w
```
No worries. That I like perl puts me in a tiny minority ... I vouch for it having worked the one time I ran it after writing it. No need to ever read it again! 😎
